File: leetcode-cn/42.接雨水.py
#
# @lc app=leetcode.cn id=42 lang=python3
#
# [42] 接雨水
#

# @lc code=start
from typing import List


class Solution:
    def trap(self, height: List[int]) -> int:
        stack = []
        ans = 0

        for i, h in enumerate(height):
            while stack and height[stack[-1]] < h:
                button_index = stack.pop(-1)

                if not stack:
                    break

                width = i - stack[-1] - 1

                hh = min(h, height[stack[-1]]) - height[button_index]

                ans += hh * width

            stack.append(i)

        return ans


# @lc code=end

File: leetcode-cn/C++/42.接雨水.cpp
/*
 * @lc app=leetcode.cn id=42 lang=cpp
 *
 * [42] 接雨水
 */

// @lc code=start
#include <vector>
#include <stack>
using namespace std;

class Solution
{
public:
    int trap(vector<int> &height)
    {
        stack<int> st;
        int ans = 0;

        for (int i = 0; i < height.size(); i++)
        {
            while (!st.empty() && height[st.top()] < height[i])
            {
                int button = st.top();
                st.pop();

                if (st.empty())
                {
                    break;
                }

                int left = st.top();

                int w = i - left - 1;
                int h = min(height[i], height[left]) - height[button];

                ans += w * h;
            }

            st.push(i);
        }

        return ans;
    }
};
// @lc code=end
